client/src/main/java/org/elasticsearch/client/RequestLogger.java

```diff
@@ -0,0 +1,138 @@
+import org.apache.http.entity.BufferedHttpEntity;
+import org.apache.http.entity.ContentType;
+import org.apache.http.util.EntityUtils;
+
+import java.io.BufferedReader;
+import java.io.IOException;
+import java.io.InputStreamReader;
+import java.nio.charset.Charset;
+import java.nio.charset.StandardCharsets;
+
+/**
+ * Helper class that exposes static methods to unify the way requests are logged.
+ * Includes trace logging to log complete requests and responses in curl format.
+ */
+final class RequestLogger {
+
+    private static final Log tracer = LogFactory.getLog("tracer");
+
+    private RequestLogger() {
+    }
+
+    /**
+     * Logs a request that yielded a response
+     */
+    static void log(Log logger, String message, HttpUriRequest request, HttpHost host, HttpResponse httpResponse) {
+        logger.debug(message + " [" + request.getMethod() + " " + host + request.getRequestLine().getUri() +
+                "] [" + httpResponse.getStatusLine() + "]");
+
+        if (tracer.isTraceEnabled()) {
+            String requestLine;
+            try {
+                requestLine = buildTraceRequest(request, host);
+            } catch(IOException e) {
+                requestLine = "";
+                tracer.trace("error while reading request for trace purposes", e);
+            }
+            String responseLine;
+            try {
+                responseLine = buildTraceResponse(httpResponse);
+            } catch(IOException e) {
+                responseLine = "";
+                tracer.trace("error while reading response for trace purposes", e);
+            }
+            tracer.trace(requestLine + '\n' + responseLine);
+        }
+    }
+
+    /**
+     * Logs a request that failed
+     */
+    static void log(Log logger, String message, HttpUriRequest request, HttpHost host, IOException e) {
+        logger.debug(message + " [" + request.getMethod() + " " + host + request.getRequestLine().getUri() + "]", e);
+        if (logger.isTraceEnabled()) {
+            String traceRequest;
+            try {
+                traceRequest = buildTraceRequest(request, host);
+            } catch (IOException e1) {
+                tracer.trace("error while reading request for trace purposes", e);
+                traceRequest = "";
+            }
+            tracer.trace(traceRequest);
+        }
+    }
+
+    /**
+     * Creates curl output for given request
+     */
+    static String buildTraceRequest(HttpUriRequest request, HttpHost host) throws IOException {
+        String requestLine = "curl -iX " + request.getMethod() + " '" + host + request.getRequestLine().getUri() + "'";
+        if (request instanceof  HttpEntityEnclosingRequest) {
+            HttpEntityEnclosingRequest enclosingRequest = (HttpEntityEnclosingRequest) request;
+            if (enclosingRequest.getEntity() != null) {
+                requestLine += " -d '";
+                HttpEntity entity = new BufferedHttpEntity(enclosingRequest.getEntity());
```

Comment:
I added an if (entity.isRepeatable == false) plus added tests for both types of entities, to make sure that logging doesn't break things.